{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser as sfp\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import defaultdict\n",
    "from bs4 import BeautifulSoup\n",
    "import seaborn as sns\n",
    "import pysam, os, glob, skbio, math, random\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cmdir(path):\n",
    "    if not os.path.isdir(path):\n",
    "        os.mkdir(path)\n",
    "\n",
    "def sbatch(name, cpus, cmd):\n",
    "    return \"sbatch -J %s -p serc -t 1- -c %d --mem %dG --wrap '%s'\" %(name, cpus, cpus*8, cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "rootdir = \"/scratch/groups/dekas/OC17_metagenomics_2022/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## contig taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "udir = rootdir + \"protein/ureC/contig_taxonomy/\"\n",
    "cmdir(udir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get/drep contigs across samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cmdir(udir + \"real_contigs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### full set\n",
    "extracted = pd.read_csv(rootdir + \"urec-oc1703a-malaspina/data/urec_filter/urec_database_details.csv\")\n",
    "extracted[\"sample\"] = extracted[\"seq_names\"].apply(lambda x: x.split(\"_sens_\")[0])\n",
    "\n",
    "for sample in extracted.query(\"urec==True\")[\"sample\"].unique():\n",
    "    \n",
    "    # find assembly\n",
    "    path = rootdir + \"assembly/clean/%s_sens_min1000.contigs.fa\" %(sample)\n",
    "\n",
    "    # get assembly\n",
    "    tmp = {}\n",
    "    for record in sfp(open(path)):\n",
    "        tmp[record[0].split(\" \")[0]] = str(record[1])\n",
    "            \n",
    "    scaffolds = list(extracted[extracted[\"sample\"]==sample].query(\"urec==True\")[\"seq_names\"].unique())\n",
    "    \n",
    "    for scaffold in scaffolds:\n",
    "        with open(udir + \"real_contigs/\" + scaffold + \".fna\", \"w\") as out:\n",
    "             out.write(\">%s\\n%s\\n\" %(scaffold, tmp[scaffold]))\n",
    "                \n",
    "tmp = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dRep compare -g /scratch/groups/dekas/OC17_metagenomics_2022/protein/ureC/contig_taxonomy/real_contigs/* -p 20 -pa 0.80 -sa 0.95 --clusterAlg single /scratch/groups/dekas/OC17_metagenomics_2022/protein/ureC/contig_taxonomy/drep\n"
     ]
    }
   ],
   "source": [
    "call = \"dRep compare -g %s -p 20 -pa 0.80 -sa 0.95 --clusterAlg single %s\" %(udir + \"real_contigs/*\", udir + \"drep\")\n",
    "print(call)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "341"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cclust = pd.read_csv(udir + \"drep/data_tables/Cdb.csv\")\n",
    "len(cclust[\"secondary_cluster\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### annotate proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cmdir(udir + \"diamond\")\n",
    "cmdir(udir + \"diamond/parts\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df = extracted.query(\"urec==True\")\n",
    "col = \"seq_names\"\n",
    "\n",
    "with open(udir + \"diamond/real_scaffold_proteins.faa\", \"w\") as out:\n",
    "    \n",
    "    for sample in df[\"sample\"].unique():\n",
    "\n",
    "        table = df[df[\"sample\"]==sample]\n",
    "        ppath = glob.glob(rootdir + \"prodigal/\" + sample + \"_sens_min1000.contigs.faa\")[0]\n",
    "        tmp = {record[0].split(\" \")[0]: str(record[1]) \\\n",
    "                   for record in sfp(open(ppath))}\n",
    "\n",
    "        tmp2 = {}\n",
    "        for key in tmp.keys():\n",
    "            scaf = \"_\".join(key.split(\"_\")[:-1])\n",
    "            if scaf not in tmp2:\n",
    "                tmp2[scaf] = [key]\n",
    "            else: tmp2[scaf].append(key)\n",
    "\n",
    "        for scaf in table[col].unique():\n",
    "            for item in tmp2[scaf]:\n",
    "                out.write(\">%s\\n%s\\n\" %(item, tmp[item]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# split and launch diamond\n",
    "records = [r for r in sfp(open(udir + \"diamond/real_scaffold_proteins.faa\"))]\n",
    "wrapper = open(udir + \"diamond/diamond.sh\", \"w\")\n",
    "\n",
    "n = math.ceil(len(records)/30)\n",
    "for i in range(0, len(records),n):\n",
    "    \n",
    "    # write partial faa\n",
    "    with open(udir + \"diamond/parts/part\" + str(int(i/n)+1) + \".faa\", \"w\") as chunk:\n",
    "        for record in records[i:i + n]:\n",
    "            chunk.write(\">%s\\n%s\\n\" %(record[0], str(record[1])))\n",
    "    \n",
    "    # generate diamond call\n",
    "    call = \"diamond blastp -d /$OAK/db/uniref100.dmnd -q %s -o %s --threads 20 -b8 -c1\" \\\n",
    "            %(udir + \"diamond/parts/part\" + str(int(i/n)+1) + \".faa\",\n",
    "              udir + \"diamond/parts/part\" + str(int(i/n)+1) + \".txt\")\n",
    "    cmd = sbatch(\"dmnd\", 20, call)\n",
    "    wrapper.write(cmd + \"\\n\")\n",
    "\n",
    "wrapper.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>qlen</th>\n",
       "      <th>qcov</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31666</th>\n",
       "      <td>OC1703_4500m_2000m_sens_contig_945622_21</td>\n",
       "      <td>UniRef100_A0A5E6R3Q4</td>\n",
       "      <td>98.1</td>\n",
       "      <td>4332.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4332.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4332.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8429.0</td>\n",
       "      <td>4333</td>\n",
       "      <td>0.999538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31871</th>\n",
       "      <td>OC1703_3000m_500m_sens_contig_2403464_31</td>\n",
       "      <td>UniRef100_UPI000C88DF8D</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2881.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2881.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2881.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5744.0</td>\n",
       "      <td>2882</td>\n",
       "      <td>0.999306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4029</th>\n",
       "      <td>OC1703_3000m_50m_sens_contig_3749701_33</td>\n",
       "      <td>UniRef100_UPI000C88DF8D</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2881.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2881.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2881.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5744.0</td>\n",
       "      <td>2882</td>\n",
       "      <td>0.999306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36131</th>\n",
       "      <td>OC1703_3000m_3000m_sens_contig_271068_2</td>\n",
       "      <td>UniRef100_UPI0015E28905</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2313.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2313.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2313.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4561.0</td>\n",
       "      <td>2314</td>\n",
       "      <td>0.999136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12929</th>\n",
       "      <td>OC1703_4500m_2000m_sens_contig_2218076_17</td>\n",
       "      <td>UniRef100_UPI000BA31F2C</td>\n",
       "      <td>98.8</td>\n",
       "      <td>2365.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2365.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2365.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4555.0</td>\n",
       "      <td>2366</td>\n",
       "      <td>0.999155</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          qseqid                   sseqid  \\\n",
       "31666   OC1703_4500m_2000m_sens_contig_945622_21     UniRef100_A0A5E6R3Q4   \n",
       "31871   OC1703_3000m_500m_sens_contig_2403464_31  UniRef100_UPI000C88DF8D   \n",
       "4029     OC1703_3000m_50m_sens_contig_3749701_33  UniRef100_UPI000C88DF8D   \n",
       "36131    OC1703_3000m_3000m_sens_contig_271068_2  UniRef100_UPI0015E28905   \n",
       "12929  OC1703_4500m_2000m_sens_contig_2218076_17  UniRef100_UPI000BA31F2C   \n",
       "\n",
       "       pident  length  mismatch  gapopen  qstart    qend  sstart    send  \\\n",
       "31666    98.1  4332.0      81.0      0.0     1.0  4332.0     1.0  4332.0   \n",
       "31871   100.0  2881.0       0.0      0.0     1.0  2881.0     1.0  2881.0   \n",
       "4029    100.0  2881.0       0.0      0.0     1.0  2881.0     1.0  2881.0   \n",
       "36131   100.0  2313.0       0.0      0.0     1.0  2313.0     1.0  2313.0   \n",
       "12929    98.8  2365.0      28.0      0.0     1.0  2365.0     1.0  2365.0   \n",
       "\n",
       "       evalue  bitscore  qlen      qcov  \n",
       "31666     0.0    8429.0  4333  0.999538  \n",
       "31871     0.0    5744.0  2882  0.999306  \n",
       "4029      0.0    5744.0  2882  0.999306  \n",
       "36131     0.0    4561.0  2314  0.999136  \n",
       "12929     0.0    4555.0  2366  0.999155  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# concatenate + collect taxonomy\n",
    "dmnd = pd.concat([skbio.io.read(item, format=\"blast+6\", into=pd.DataFrame, default_columns=True) \\\n",
    "                      for item in glob.glob(udir + \"diamond/parts/part*.txt\")])\n",
    "# compute coverage\n",
    "faalens = {record[0].split(\" \")[0]: len(record[1]) for record \\\n",
    "           in sfp(open(udir + \"diamond/real_scaffold_proteins.faa\"))}\n",
    "dmnd[\"qlen\"] = dmnd[\"qseqid\"].map(faalens)\n",
    "dmnd[\"qcov\"] = dmnd.apply(lambda x: (x[\"qend\"]-x[\"qstart\"])/x[\"qlen\"], axis=1)\n",
    "# choose best hits for each\n",
    "dmnd = dmnd.sort_values([\"bitscore\", \"qcov\"], ascending=[False,False]).drop_duplicates(\"qseqid\")\n",
    "# filter for min cov /eval\n",
    "dmnd = dmnd[(dmnd[\"evalue\"]<1e-20) & (dmnd[\"qcov\"]>0.70)]\n",
    "dmnd.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# map to uniref ids\n",
    "with open(udir + \"diamond/uniref_ids.txt\", \"w\") as out:\n",
    "    for uid in dmnd[\"sseqid\"].unique():\n",
    "        out.write(uid.split(\"_\")[1] + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "N.B. need to do two searches, one for uniref and one for uniparc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>entry</th>\n",
       "      <th>organism</th>\n",
       "      <th>organism_id</th>\n",
       "      <th>taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>906</th>\n",
       "      <td>UPI000732158F</td>\n",
       "      <td>UPI000732158F</td>\n",
       "      <td>Pseudomonas fluorescens; Pseudomonas fluorescens</td>\n",
       "      <td>294; 294</td>\n",
       "      <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907</th>\n",
       "      <td>UPI001B3439E0</td>\n",
       "      <td>UPI001B3439E0</td>\n",
       "      <td>Pseudomonas protegens</td>\n",
       "      <td>380021</td>\n",
       "      <td>380021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>908</th>\n",
       "      <td>UPI001BB016DC</td>\n",
       "      <td>UPI001BB016DC</td>\n",
       "      <td>Marinomonas sp. CT5; Marinomonas sp. CT5</td>\n",
       "      <td>2066133; 2066133</td>\n",
       "      <td>2066133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>909</th>\n",
       "      <td>UPI0002319E8B</td>\n",
       "      <td>UPI0002319E8B</td>\n",
       "      <td>Pseudoalteromonas sp. BSi20480; unclassified P...</td>\n",
       "      <td>386428; 194690; 386428; 386428; 386428; 386428...</td>\n",
       "      <td>386428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910</th>\n",
       "      <td>UPI0021F8827D</td>\n",
       "      <td>UPI0021F8827D</td>\n",
       "      <td>Rheinheimera sp. MM224; Rheinheimera sp.</td>\n",
       "      <td>3019969; 1869214</td>\n",
       "      <td>1869214</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              from          entry  \\\n",
       "906  UPI000732158F  UPI000732158F   \n",
       "907  UPI001B3439E0  UPI001B3439E0   \n",
       "908  UPI001BB016DC  UPI001BB016DC   \n",
       "909  UPI0002319E8B  UPI0002319E8B   \n",
       "910  UPI0021F8827D  UPI0021F8827D   \n",
       "\n",
       "                                              organism  \\\n",
       "906   Pseudomonas fluorescens; Pseudomonas fluorescens   \n",
       "907                              Pseudomonas protegens   \n",
       "908           Marinomonas sp. CT5; Marinomonas sp. CT5   \n",
       "909  Pseudoalteromonas sp. BSi20480; unclassified P...   \n",
       "910           Rheinheimera sp. MM224; Rheinheimera sp.   \n",
       "\n",
       "                                           organism_id    taxid  \n",
       "906                                           294; 294      294  \n",
       "907                                             380021   380021  \n",
       "908                                   2066133; 2066133  2066133  \n",
       "909  386428; 194690; 386428; 386428; 386428; 386428...   386428  \n",
       "910                                   3019969; 1869214  1869214  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniref = pd.read_csv(udir + \"diamond/uniref_idmapping.tsv\", sep=\"\\t\")\n",
    "uniref.columns = [\"from\", \"entry\", \"organism\", \"organism_id\"]\n",
    "uniref[\"organism_id\"] = uniref[\"organism_id\"].apply(lambda x: str(x))\n",
    "uniparc = pd.read_csv(udir + \"diamond/uniparc_idmapping.tsv\", sep=\"\\t\")\n",
    "uniparc.columns = [\"from\", \"entry\", \"organism\", \"organism_id\"]\n",
    "uni = pd.concat([uniref, uniparc]).fillna(\"None\")\n",
    "uni[\"taxid\"] = uni[\"organism_id\"].apply(lambda x: stats.mode(x.split(\"; \"))[0][0])\n",
    "uni.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with open(udir + \"diamond/ncbi_tax.txt\", \"w\") as out:\n",
    "    for tax in uni[\"taxid\"].unique():\n",
    "        out.write(tax + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lineage_info = defaultdict(list)\n",
    "\n",
    "for xml in glob.glob(udir + \"/diamond/*xml\"):\n",
    "    \n",
    "    for block in BeautifulSoup(open(xml), \"xml\").findAll(\"Taxon\"):\n",
    "        \n",
    "        lineage, phylum = \"None\", \"None\"\n",
    "\n",
    "        if block.find(\"Lineage\"):\n",
    "            lineage = block.find(\"Lineage\").string\n",
    "\n",
    "        for level in block.findAll(\"Taxon\"):\n",
    "            if level.find(\"Rank\").string==\"phylum\":\n",
    "                phylum = level.find(\"ScientificName\").string\n",
    "\n",
    "        lineage_info[\"taxid\"].append(block.find(\"TaxId\").string)\n",
    "        lineage_info[\"lineage\"].append(lineage)\n",
    "        lineage_info[\"phylum\"].append(phylum)\n",
    "\n",
    "lineage_df = pd.DataFrame(lineage_info).query(\"lineage!='None'\")\n",
    "uni = uni.merge(lineage_df, how=\"left\", on=\"taxid\")\n",
    "uni.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>qlen</th>\n",
       "      <th>qcov</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>entry</th>\n",
       "      <th>phylum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OC1703_3000m_500m_sens_contig_2403464_31</td>\n",
       "      <td>UniRef100_UPI000C88DF8D</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2881.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2881.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2881.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5744.0</td>\n",
       "      <td>2882</td>\n",
       "      <td>0.999306</td>\n",
       "      <td>OC1703_3000m_500m_sens_contig_2403464</td>\n",
       "      <td>UPI000C88DF8D</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OC1703_3000m_50m_sens_contig_3749701_33</td>\n",
       "      <td>UniRef100_UPI000C88DF8D</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2881.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2881.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2881.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5744.0</td>\n",
       "      <td>2882</td>\n",
       "      <td>0.999306</td>\n",
       "      <td>OC1703_3000m_50m_sens_contig_3749701</td>\n",
       "      <td>UPI000C88DF8D</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OC1703_3000m_3000m_sens_contig_271068_2</td>\n",
       "      <td>UniRef100_UPI0015E28905</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2313.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2313.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2313.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4561.0</td>\n",
       "      <td>2314</td>\n",
       "      <td>0.999136</td>\n",
       "      <td>OC1703_3000m_3000m_sens_contig_271068</td>\n",
       "      <td>UPI0015E28905</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OC1703_4500m_2000m_sens_contig_2218076_17</td>\n",
       "      <td>UniRef100_UPI000BA31F2C</td>\n",
       "      <td>98.8</td>\n",
       "      <td>2365.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2365.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2365.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4555.0</td>\n",
       "      <td>2366</td>\n",
       "      <td>0.999155</td>\n",
       "      <td>OC1703_4500m_2000m_sens_contig_2218076</td>\n",
       "      <td>UPI000BA31F2C</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OC1703_3000m_1000m_sens_contig_2825716_12</td>\n",
       "      <td>UniRef100_UPI0015E28905</td>\n",
       "      <td>99.9</td>\n",
       "      <td>2313.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2313.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2313.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4554.0</td>\n",
       "      <td>2314</td>\n",
       "      <td>0.999136</td>\n",
       "      <td>OC1703_3000m_1000m_sens_contig_2825716</td>\n",
       "      <td>UPI0015E28905</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      qseqid                   sseqid  pident  \\\n",
       "1   OC1703_3000m_500m_sens_contig_2403464_31  UniRef100_UPI000C88DF8D   100.0   \n",
       "2    OC1703_3000m_50m_sens_contig_3749701_33  UniRef100_UPI000C88DF8D   100.0   \n",
       "3    OC1703_3000m_3000m_sens_contig_271068_2  UniRef100_UPI0015E28905   100.0   \n",
       "4  OC1703_4500m_2000m_sens_contig_2218076_17  UniRef100_UPI000BA31F2C    98.8   \n",
       "5  OC1703_3000m_1000m_sens_contig_2825716_12  UniRef100_UPI0015E28905    99.9   \n",
       "\n",
       "   length  mismatch  gapopen  qstart    qend  sstart    send  evalue  \\\n",
       "1  2881.0       0.0      0.0     1.0  2881.0     1.0  2881.0     0.0   \n",
       "2  2881.0       0.0      0.0     1.0  2881.0     1.0  2881.0     0.0   \n",
       "3  2313.0       0.0      0.0     1.0  2313.0     1.0  2313.0     0.0   \n",
       "4  2365.0      28.0      0.0     1.0  2365.0     1.0  2365.0     0.0   \n",
       "5  2313.0       3.0      0.0     1.0  2313.0     1.0  2313.0     0.0   \n",
       "\n",
       "   bitscore  qlen      qcov                                scaffold  \\\n",
       "1    5744.0  2882  0.999306   OC1703_3000m_500m_sens_contig_2403464   \n",
       "2    5744.0  2882  0.999306    OC1703_3000m_50m_sens_contig_3749701   \n",
       "3    4561.0  2314  0.999136   OC1703_3000m_3000m_sens_contig_271068   \n",
       "4    4555.0  2366  0.999155  OC1703_4500m_2000m_sens_contig_2218076   \n",
       "5    4554.0  2314  0.999136  OC1703_3000m_1000m_sens_contig_2825716   \n",
       "\n",
       "           entry          phylum  \n",
       "1  UPI000C88DF8D  Pseudomonadota  \n",
       "2  UPI000C88DF8D  Pseudomonadota  \n",
       "3  UPI0015E28905  Pseudomonadota  \n",
       "4  UPI000BA31F2C  Pseudomonadota  \n",
       "5  UPI0015E28905  Pseudomonadota  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add back in\n",
    "dmnd[\"scaffold\"] = dmnd[\"qseqid\"].apply(lambda x: \"_\".join(x.split(\"_\")[:-1]))\n",
    "dmnd[\"entry\"] = dmnd[\"sseqid\"].apply(lambda x: x.split(\"_\")[1])\n",
    "dmnd = dmnd.merge(uni[[\"entry\", \"phylum\"]], how=\"left\").fillna(\"None\")\n",
    "# fix above\n",
    "dmnd = dmnd[dmnd[\"scaffold\"].isin(extracted.query(\"urec==True\")[\"seq_names\"].unique())]\n",
    "dmnd.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### reconcile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5690"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orf_counts = {}\n",
    "\n",
    "for record in sfp(open(udir + \"diamond/real_scaffold_proteins.faa\")):\n",
    "    scaf = \"_\".join(record[0].split(\"_\")[:-1])\n",
    "    if scaf not in orf_counts:\n",
    "        orf_counts[scaf] = 1\n",
    "    else: orf_counts[scaf] +=1\n",
    "    \n",
    "len(orf_counts.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>prot_phylum_winner</th>\n",
       "      <th>prot_phylum_winner_perc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1795</th>\n",
       "      <td>OC1703_4500m_1000m_sens_contig_1902572</td>\n",
       "      <td>Nitrospinota</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1796</th>\n",
       "      <td>OC1703_4500m_3000m_sens_contig_3530469</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1797</th>\n",
       "      <td>OC1703_3000m_150m_sens_contig_3866675</td>\n",
       "      <td>None</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1798</th>\n",
       "      <td>OC1703_4500m_1000m_sens_contig_3043723</td>\n",
       "      <td>Verrucomicrobiota</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1799</th>\n",
       "      <td>OC1703_4500m_500m_sens_contig_5062989</td>\n",
       "      <td>None</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    scaffold prot_phylum_winner  \\\n",
       "1795  OC1703_4500m_1000m_sens_contig_1902572       Nitrospinota   \n",
       "1796  OC1703_4500m_3000m_sens_contig_3530469     Pseudomonadota   \n",
       "1797   OC1703_3000m_150m_sens_contig_3866675               None   \n",
       "1798  OC1703_4500m_1000m_sens_contig_3043723  Verrucomicrobiota   \n",
       "1799   OC1703_4500m_500m_sens_contig_5062989               None   \n",
       "\n",
       "      prot_phylum_winner_perc  \n",
       "1795                 0.333333  \n",
       "1796                 1.000000  \n",
       "1797                 0.500000  \n",
       "1798                 0.666667  \n",
       "1799                 0.500000  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tax_info = defaultdict(list)\n",
    "\n",
    "for scaf in dmnd[\"scaffold\"].unique():\n",
    "    \n",
    "    subtable = dmnd[dmnd[\"scaffold\"]==scaf].groupby([\"scaffold\", \"phylum\"], \\\n",
    "        as_index=False).aggregate({\"qseqid\":\"count\"}).sort_values(\"qseqid\", ascending=False)\n",
    "    subtable[\"total_orfs\"] = orf_counts[subtable[\"scaffold\"].iloc[0]]\n",
    "    subtable[\"perc_orfs\"] = subtable.apply(lambda x: x[\"qseqid\"]/x[\"total_orfs\"], axis=1)\n",
    "    sorted_table = subtable.sort_values(\"perc_orfs\", ascending=False)\n",
    "    tax_info[\"scaffold\"].append(sorted_table[\"scaffold\"].iloc[0])\n",
    "    tax_info[\"prot_phylum_winner\"].append(sorted_table[\"phylum\"].iloc[0])\n",
    "    tax_info[\"prot_phylum_winner_perc\"].append(sorted_table[\"perc_orfs\"].iloc[0])\n",
    "\n",
    "tax_info_df = pd.DataFrame(tax_info)\n",
    "tax_info_df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### curate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tmerge = tax_info_df\n",
    "\n",
    "tmp = {}\n",
    "for contig in glob.glob(udir + \"real_contigs/*\"):\n",
    "    for record in sfp(open(contig)):\n",
    "        tmp[record[0]]=len(record[1])\n",
    "tmerge[\"scaffold_len\"] = tmerge[\"scaffold\"].map(tmp)\n",
    "tmp = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "binned = pd.read_csv(rootdir + \"protein/ureC/urec_binned_contigs.tsv\", sep=\"\\t\")\n",
    "tmerge = tmerge.merge(binned[[\"scaffold\", \"classification\"]], how=\"left\").fillna(\"None\")\n",
    "tmerge = tmerge.rename(columns={\"classification\":\"bin_classification\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cclust[\"scaffold\"] = cclust[\"genome\"].apply(lambda x: x.replace(\".fna\", \"\"))\n",
    "tmerge = tmerge.merge(cclust[[\"scaffold\", \"primary_cluster\",\"secondary_cluster\"]], how=\"left\").fillna(\"not clustered\")\n",
    "tmerge[[\"scaffold\", \"scaffold_len\", \"primary_cluster\", \"secondary_cluster\", \\\n",
    "        \"prot_phylum_winner\", \"prot_phylum_winner_perc\", \"bin_classification\"]].sort_values([\"primary_cluster\", \\\n",
    "        \"secondary_cluster\", \"scaffold_len\"], ascending=[True,True,False]).to_csv(udir + \"diamond/to_curate.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>scaffold_len</th>\n",
       "      <th>primary_cluster</th>\n",
       "      <th>secondary_cluster</th>\n",
       "      <th>prot_phylum_winner</th>\n",
       "      <th>prot_phylum_winner_perc</th>\n",
       "      <th>bin_classification</th>\n",
       "      <th>manual</th>\n",
       "      <th>reconciled_taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OC1703_4500m_500m_sens_contig_3438907</td>\n",
       "      <td>11706</td>\n",
       "      <td>1</td>\n",
       "      <td>1_0</td>\n",
       "      <td>Chloroflexota</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Chloroflexota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OC1703_3000m_1000m_sens_contig_1343963</td>\n",
       "      <td>290156</td>\n",
       "      <td>2</td>\n",
       "      <td>2_1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>0.950758</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OC1703_3000m_3000m_sens_contig_701093</td>\n",
       "      <td>133857</td>\n",
       "      <td>2</td>\n",
       "      <td>2_1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>0.896825</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OC1703_4500m_2000m_sens_contig_1752643</td>\n",
       "      <td>19664</td>\n",
       "      <td>2</td>\n",
       "      <td>2_1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>0.956522</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OC1703_3000m_50m_sens_contig_1313614</td>\n",
       "      <td>9557</td>\n",
       "      <td>3</td>\n",
       "      <td>3_1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OC1703_4500m_50m_sens_contig_6005247</td>\n",
       "      <td>5078</td>\n",
       "      <td>3</td>\n",
       "      <td>3_1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>OC1703_3000m_50m_sens_contig_1315830</td>\n",
       "      <td>123315</td>\n",
       "      <td>7</td>\n",
       "      <td>7_1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>0.958678</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>OC1703_3000m_500m_sens_contig_5191787</td>\n",
       "      <td>123314</td>\n",
       "      <td>7</td>\n",
       "      <td>7_1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>0.958678</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>OC1703_3000m_1000m_sens_contig_3276135</td>\n",
       "      <td>123314</td>\n",
       "      <td>7</td>\n",
       "      <td>7_1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>0.958678</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>OC1703_3000m_150m_sens_contig_1823447</td>\n",
       "      <td>35336</td>\n",
       "      <td>7</td>\n",
       "      <td>7_1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteobacteria</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  scaffold  scaffold_len primary_cluster  \\\n",
       "0    OC1703_4500m_500m_sens_contig_3438907         11706               1   \n",
       "1   OC1703_3000m_1000m_sens_contig_1343963        290156               2   \n",
       "2    OC1703_3000m_3000m_sens_contig_701093        133857               2   \n",
       "3   OC1703_4500m_2000m_sens_contig_1752643         19664               2   \n",
       "4     OC1703_3000m_50m_sens_contig_1313614          9557               3   \n",
       "5     OC1703_4500m_50m_sens_contig_6005247          5078               3   \n",
       "8     OC1703_3000m_50m_sens_contig_1315830        123315               7   \n",
       "9    OC1703_3000m_500m_sens_contig_5191787        123314               7   \n",
       "10  OC1703_3000m_1000m_sens_contig_3276135        123314               7   \n",
       "11   OC1703_3000m_150m_sens_contig_1823447         35336               7   \n",
       "\n",
       "   secondary_cluster prot_phylum_winner  prot_phylum_winner_perc  \\\n",
       "0                1_0      Chloroflexota                 0.833333   \n",
       "1                2_1     Proteobacteria                 0.950758   \n",
       "2                2_1     Proteobacteria                 0.896825   \n",
       "3                2_1     Proteobacteria                 0.956522   \n",
       "4                3_1     Proteobacteria                 1.000000   \n",
       "5                3_1     Proteobacteria                 1.000000   \n",
       "8                7_1     Proteobacteria                 0.958678   \n",
       "9                7_1     Proteobacteria                 0.958678   \n",
       "10               7_1     Proteobacteria                 0.958678   \n",
       "11               7_1     Proteobacteria                 0.948718   \n",
       "\n",
       "                                   bin_classification manual  \\\n",
       "0                                                None   None   \n",
       "1   d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   None   \n",
       "2   d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   None   \n",
       "3                                                None   None   \n",
       "4                                                None   None   \n",
       "5                                                None   None   \n",
       "8                                                None   None   \n",
       "9                                                None   None   \n",
       "10                                               None   None   \n",
       "11                                               None   None   \n",
       "\n",
       "   reconciled_taxonomy  \n",
       "0        Chloroflexota  \n",
       "1       Proteobacteria  \n",
       "2       Proteobacteria  \n",
       "3       Proteobacteria  \n",
       "4       Proteobacteria  \n",
       "5       Proteobacteria  \n",
       "8       Proteobacteria  \n",
       "9       Proteobacteria  \n",
       "10      Proteobacteria  \n",
       "11      Proteobacteria  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read back in\n",
    "def reconcile_taxonomy(row):\n",
    "    if row[\"manual\"]!='None':\n",
    "        return row[\"manual\"]\n",
    "    elif row[\"prot_phylum_winner\"] == 'None':\n",
    "        return \"unclassified\"\n",
    "    else: return row[\"prot_phylum_winner\"]\n",
    "        \n",
    "curated = pd.read_csv(udir + \"diamond/curated.csv\").fillna(\"None\")\n",
    "curated[\"reconciled_taxonomy\"] = curated.apply(reconcile_taxonomy, axis=1)\n",
    "curated.query(\"manual=='None'\").head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>scaffold_len</th>\n",
       "      <th>primary_cluster</th>\n",
       "      <th>secondary_cluster</th>\n",
       "      <th>prot_phylum_winner</th>\n",
       "      <th>prot_phylum_winner_perc</th>\n",
       "      <th>bin_classification</th>\n",
       "      <th>manual</th>\n",
       "      <th>reconciled_taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>OC1703_3000m_50m_sens_contig_2514702</td>\n",
       "      <td>9169</td>\n",
       "      <td>6</td>\n",
       "      <td>6_1</td>\n",
       "      <td>Chlorophyta</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>None</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>eukaryote</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>OC1703_4500m_50m_sens_contig_6461713</td>\n",
       "      <td>8500</td>\n",
       "      <td>6</td>\n",
       "      <td>6_1</td>\n",
       "      <td>Chlorophyta</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>None</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>eukaryote</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>OC1703_3000m_50m_sens_contig_22604</td>\n",
       "      <td>4402</td>\n",
       "      <td>65</td>\n",
       "      <td>65_1</td>\n",
       "      <td>Chlorophyta</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>None</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>eukaryote</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>OC1703_4500m_50m_sens_contig_4421038</td>\n",
       "      <td>4060</td>\n",
       "      <td>65</td>\n",
       "      <td>65_1</td>\n",
       "      <td>Chlorophyta</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>None</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>eukaryote</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>662</th>\n",
       "      <td>OC1703_4500m_50m_sens_contig_2626437</td>\n",
       "      <td>3347</td>\n",
       "      <td>176</td>\n",
       "      <td>176_0</td>\n",
       "      <td>None</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>None</td>\n",
       "      <td>eukaryote</td>\n",
       "      <td>eukaryote</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 scaffold  scaffold_len primary_cluster  \\\n",
       "6    OC1703_3000m_50m_sens_contig_2514702          9169               6   \n",
       "7    OC1703_4500m_50m_sens_contig_6461713          8500               6   \n",
       "115    OC1703_3000m_50m_sens_contig_22604          4402              65   \n",
       "116  OC1703_4500m_50m_sens_contig_4421038          4060              65   \n",
       "662  OC1703_4500m_50m_sens_contig_2626437          3347             176   \n",
       "\n",
       "    secondary_cluster prot_phylum_winner  prot_phylum_winner_perc  \\\n",
       "6                 6_1        Chlorophyta                 0.833333   \n",
       "7                 6_1        Chlorophyta                 1.000000   \n",
       "115              65_1        Chlorophyta                 0.500000   \n",
       "116              65_1        Chlorophyta                 0.500000   \n",
       "662             176_0               None                 1.000000   \n",
       "\n",
       "    bin_classification     manual reconciled_taxonomy  \n",
       "6                 None  eukaryote           eukaryote  \n",
       "7                 None  eukaryote           eukaryote  \n",
       "115               None  eukaryote           eukaryote  \n",
       "116               None  eukaryote           eukaryote  \n",
       "662               None  eukaryote           eukaryote  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how many euks?\n",
    "curated.query(\"manual=='eukaryote'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/users/ajaffe/.local/lib/python3.6/site-packages/seaborn/_decorators.py:43: FutureWarning: Pass the following variables as keyword args: x, y. From version 0.12, the only valid positional argument will be `data`, and passing other arguments without an explicit keyword will result in an error or misinterpretation.\n",
      "  FutureWarning\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEGCAYAAACKB4k+AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90\nbGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsT\nAAALEwEAmpwYAAAXDUlEQVR4nO3da5RlZX3n8e9PQEBEUakwLY1pYhDGS9JgBSQQRRkV0QSMRnGN\niohDxgVeVlwqmFnxkhBNjDKSRGa1gDTeEEGXBEkUEQUSAauhuYN2uAy0SBcICF4Ygf+8OE9vyqa6\n6nRT55zu6u9nrb3O3s9+9t7/0y/6V/tynp2qQpIkgMeNugBJ0obDUJAkdQwFSVLHUJAkdQwFSVJn\n81EX8Fhsv/32tWjRolGXIUkblWXLlt1ZVWPTrduoQ2HRokVMTEyMugxJ2qgkuWVt67x8JEnqGAqS\npI6hIEnqGAqSpI6hIEnqGAqSpI6hIEnqGAqSpI6hIEnqbNS/aN5U/d+PPG/UJayTZ/zVVaMuQVKf\nPFOQJHUMBUlSx1CQJHUMBUlSx1CQJHUMBUlSx1CQJHUMBUlSZ2ChkGSrJJcmuSLJNUk+3NpPSXJT\nkuVtWtzak+T4JCuSXJlkj0HVJkma3iB/0fwA8JKquj/JFsBFSf61rXtvVZ2xRv9XALu0aS/ghPYp\nSRqSgZ0pVM/9bXGLNtUMmxwEnNq2uxjYLsmCQdUnSXq0gd5TSLJZkuXAKuDcqrqkrTq2XSI6LsmW\nrW1H4NYpm9/W2tbc5xFJJpJMTE5ODrJ8SdrkDDQUquqhqloMLAT2TPJc4BhgN+APgKcC71/HfS6p\nqvGqGh8bG5vrkiVpkzaUp4+q6h7gfOCAqrq9XSJ6APgssGfrthLYacpmC1ubJGlIBvn00ViS7dr8\n1sBLgetX3ydIEuBg4Oq2yVnAm9tTSC8A7q2q2wdVnyTp0Qb59NECYGmSzeiFz+lVdXaS7yQZAwIs\nB/5n638OcCCwAvgFcNgAa5MkTWNgoVBVVwK7T9P+krX0L+DIQdUjSZqdb17TBmWff9xn1CWss39/\nx7+PugRpzjjMhSSpYyhIkjqGgiSpYyhIkjqGgiSpYyhIkjqGgiSpYyhIkjqGgiSpYyhIkjqGgiSp\nYyhIkjqGgiSpYyhIkjqGgiSpYyhIkjqGgiSpYyhIkjoDC4UkWyW5NMkVSa5J8uHWvnOSS5KsSPLl\nJI9v7Vu25RVt/aJB1SZJmt4gzxQeAF5SVb8PLAYOSPIC4O+A46rqd4G7gcNb/8OBu1v7ca2fJGmI\nBhYK1XN/W9yiTQW8BDijtS8FDm7zB7Vl2vr9k2RQ9UmSHm2g9xSSbJZkObAKOBf4T+CeqnqwdbkN\n2LHN7wjcCtDW3ws8bZp9HpFkIsnE5OTkIMuXpE3OQEOhqh6qqsXAQmBPYLc52OeSqhqvqvGxsbHH\nujtJ0hRDefqoqu4Bzgf2BrZLsnlbtRBY2eZXAjsBtPVPBu4aRn2SpJ5BPn00lmS7Nr818FLgOnrh\n8NrW7VDg623+rLZMW/+dqqpB1SdJerTNZ++y3hYAS5NsRi98Tq+qs5NcC5yW5G+Ay4GTWv+TgM8l\nWQH8FDhkgLVJkqYxsFCoqiuB3adpv5He/YU1238F/Nmg6pEkzc5fNEuSOoaCJKljKEiSOoaCJKlj\nKEiSOoaCJKljKEiSOoaCJKljKEiSOoaCJKljKEiSOoaCJKljKEiSOoaCJKljKEiSOoaCJKljKEiS\nOoaCJKkzsFBIslOS85Ncm+SaJO9q7R9KsjLJ8jYdOGWbY5KsSHJDkpcPqjZJ0vQG9o5m4EHgPVV1\nWZJtgWVJzm3rjquqf5jaOcmzgUOA5wBPB76d5FlV9dAAa5QkTTGwM4Wqur2qLmvz9wHXATvOsMlB\nwGlV9UBV3QSsAPYcVH2SpEcbyj2FJIuA3YFLWtNRSa5McnKSp7S2HYFbp2x2GzOHiCRpjg08FJI8\nETgTeHdV/Qw4AXgmsBi4HfjEOu7viCQTSSYmJyfnulxJ2qQNNBSSbEEvEL5QVV8FqKo7quqhqnoY\n+AyPXCJaCew0ZfOFre03VNWSqhqvqvGxsbFBli9Jm5xZQyHJ3yd5UpItkpyXZDLJG/vYLsBJwHVV\n9ckp7QumdHs1cHWbPws4JMmWSXYGdgEuXZcvI0l6bPp5+uhlVfW+JK8Gbgb+FLgA+Pws2+0DvAm4\nKsny1vYB4A1JFgPV9vfnAFV1TZLTgWvpPbl0pE8eSdJw9RMKq/u8EvhKVd3bOwmYWVVdBEzX8ZwZ\ntjkWOLaPmiRJA9BPKJyd5Hrgl8Dbk4wBvxpsWZKkUZj1nkJVHQ38ITBeVb8GfkHvNwWSpHlm1jOF\nJH86ZX717L1JHq6qVYMqTJI0fP1cPjoc2Bs4vy3vBywDdk7ykar63IBqkyQNWb83mv9rVd0BkGQH\n4FRgL3pPIRkKkjRP9PPjtZ1WB0KzqrX9FPj1YMqSJI1CP2cK301yNvCVtvya1rYNcM+gCpMkDV8/\noXAkvSDYpy2fCpxZVQW8eFCFSZKGb9ZQaP/5n9EmSdI8ttZQSHJRVe2b5D56Q1J0q+hlxZMGXp0k\naajWGgpVtW/73HZ45UiSRqmfUVIf9cjpdG2SpI1fP4+kPmfqQpLNgecPphxJ0iitNRSSHNPuJ/xe\nkp+16T7gDuDrQ6tQkjQ0M91T+Cjw0SQfrapjhljTY/b895466hLWybKPv3nUJUgS0N8jqcck2RH4\n7an9q+qCQRYmSRq+fkZJ/RhwCL03oq1+E1rRG/dIkjSP9POL5lcDu1bVA4MuRpI0Wv08fXQjsMWg\nC5EkjV4/Zwq/AJYnOQ/ozhaq6p0zbZRkJ3rjJO1A73LTkqr6VJKnAl8GFgE3A6+rqrvTe4PPp4AD\n2zHfUlWXrfM3kiStt35C4aw2rasHgfdU1WVJtgWWJTkXeAtwXlV9LMnRwNHA+4FXALu0aS/ghPYp\nSRqSfp4+Wprk8cCzWtMN7V3Ns213O3B7m78vyXXAjvTe77xf67YU+C69UDgIOLUNwHdxku2SLGj7\nkSQNQT/DXOwH/Aj4Z+DTwA+TvHBdDpJkEbA7cAmww5T/6H9C7/IS9ALj1imb3dba1tzXEUkmkkxM\nTk6uSxmSpFn0c6P5E8DLqupFVfVC4OXAcf0eIMkTgTOBd1fVz6aua2cFNe2Ga1FVS6pqvKrGx8bG\n1mVTSdIs+gmFLarqhtULVfVD+nwaKckW9ALhC1X11dZ8R5IFbf0Ceq/3BFgJ7DRl84WtTZI0JP2E\nwkSSE5Ps16bPABOzbdSeJjoJuK6qPjll1VnAoW3+UB4ZR+ks4M3peQFwr/cTJGm4+nn66O30Xsm5\n+hHUC+ndW5jNPsCbgKuSLG9tHwA+Bpye5HDgFuB1bd059B5HXUHvkdTD+jiGJGkO9RMKmwOfWv3X\nfpLNgC1n26iqLqL3lrbp7D9N/6IXPpKkEenn8tF5wNZTlrcGvj2YciRJo9RPKGxVVfevXmjzTxhc\nSZKkUeknFH6eZI/VC0meD/xycCVJkkaln3sK7wa+kuTH9O4R/Bfg9YMsSpI0Gv0Mc/GDJLsBu7am\nvoa5kCRtfPo5U6CFwNUDrkWSNGL93FOQJG0i1hoKSfZpn7P+JkGSND/MdKZwfPv8/jAKkSSN3kz3\nFH6dZAmwY5Lj11w525vXJEkbn5lC4VXAf6M3VPay4ZQjSRqltYZCVd0JnJbkuqq6Yog1SZJGpJ+n\nj+5K8rUkq9p0ZpKFA69MkjR0/YTCZ+m96+DpbfqX1iZJmmf6CYXfqqrPVtWDbToF8D2YkjQP9RMK\ndyZ5Y5LN2vRG4K5BFyZJGr5+QuGt9N6O9hPgduC1+FY0SZqX+hkQ7xbgT4ZQiyRpxBz7SJLUGVgo\nJDm5PcJ69ZS2DyVZmWR5mw6csu6YJCuS3JDk5YOqS5K0doM8UzgFOGCa9uOqanGbzgFI8mzgEOA5\nbZtPJ9lsgLVJkqYxaygk+V9T5vseMbWqLgB+2mf3g4DTquqBqroJWAHs2e+xJElzY6ahs9+fZG96\nTxutNhcjph6V5Mp2eekprW1H4NYpfW5rbdPVdUSSiSQTk5OTc1COJGm1mc4Urgf+DPidJBcm+Qzw\ntCS7zrDNbE4Angkspvd46yfWdQdVtaSqxqtqfGzM39BJ0lyaKRTuAT5A71LOfsCnWvvRSf5jfQ5W\nVXdU1UNV9TDwGR65RLQS2GlK14WtTZI0RDOFwsuBb9D7y/6TwF7Az6vqsKr6w/U5WJIFUxZfzSPv\nfT4LOCTJlkl2BnYBLl2fY0iS1t9MQ2d/ACDJFcDngD2AsSQXAXdX1R/PtOMkX6J3hrF9ktuADwL7\nJVkMFHAz8OftWNckOR24FngQOLKqHnpM30yStM5m/UUz8M2qmgAmkry9qvZNsv1sG1XVG6ZpPmmG\n/scCx/ZRjyRpQPoZ5uJ9Uxbf0truHFRB0nz2vRe+aNQlrLMXXfC9UZegIVqnH6/5BjZJmt8c+0iS\n1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEU\nJEkdQ0GS1DEUJEkdQ0GS1BlYKCQ5OcmqJFdPaXtqknOT/Kh9PqW1J8nxSVYkuTLJHoOqS5K0doM8\nUzgFOGCNtqOB86pqF+C8tgzwCmCXNh0BnDDAuiRJazGwUKiqC4CfrtF8ELC0zS8FDp7Sfmr1XAxs\nl2TBoGqTJE1v2PcUdqiq29v8T4Ad2vyOwK1T+t3W2h4lyRFJJpJMTE5ODq5SSdoEjexGc1UVUOux\n3ZKqGq+q8bGxsQFUJkmbrmGHwh2rLwu1z1WtfSWw05R+C1ubJGmIhh0KZwGHtvlDga9PaX9zewrp\nBcC9Uy4zSZKGZPNB7TjJl4D9gO2T3AZ8EPgYcHqSw4FbgNe17ucABwIrgF8Ahw2qLknS2g0sFKrq\nDWtZtf80fQs4clC1SJL64y+aJUkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEkd\nQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEmdgb2jeSZJbgbuAx4C\nHqyq8SRPBb4MLAJuBl5XVXePoj5J2lSN8kzhxVW1uKrG2/LRwHlVtQtwXluWJA3RhnT56CBgaZtf\nChw8ulIkadM0qlAo4FtJliU5orXtUFW3t/mfADtMt2GSI5JMJJmYnJwcRq2StMkYyT0FYN+qWpnk\nt4Bzk1w/dWVVVZKabsOqWgIsARgfH5+2jyRp/YzkTKGqVrbPVcDXgD2BO5IsAGifq0ZRmyRtyoYe\nCkm2SbLt6nngZcDVwFnAoa3bocDXh12bJG3qRnH5aAfga0lWH/+LVfVvSX4AnJ7kcOAW4HUjqE2S\nNmlDD4WquhH4/Wna7wL2H3Y9kqRHbEiPpEqSRsxQkCR1RvVIqqR56J/e8y+jLmGdHfWJPx51CRsU\nzxQkSR1DQZLUMRQkSR1DQZLUMRQkSR1DQZLUMRQkSR1DQZLUMRQkSR1DQZLUMRQkSR1DQZLUMRQk\nSR1DQZLUMRQkSR1DQZLU2eBCIckBSW5IsiLJ0aOuR5I2JRtUKCTZDPhn4BXAs4E3JHn2aKuSpE3H\nhvY6zj2BFVV1I0CS04CDgGtHWpUkAce+8bWjLmGd/OXnz1jnbVJVAyhl/SR5LXBAVb2tLb8J2Kuq\njprS5wjgiLa4K3DDEEvcHrhziMcbNr/fxms+fzfw+821366qselWbGhnCrOqqiXAklEcO8lEVY2P\n4tjD4PfbeM3n7wZ+v2HaoO4pACuBnaYsL2xtkqQh2NBC4QfALkl2TvJ44BDgrBHXJEmbjA3q8lFV\nPZjkKOCbwGbAyVV1zYjLmmokl62GyO+38ZrP3w38fkOzQd1oliSN1oZ2+UiSNEKGgiSpYyj0IcnJ\nSVYluXrUtcy1JDslOT/JtUmuSfKuUdc0l5JsleTSJFe07/fhUdc0CEk2S3J5krNHXctcS3JzkquS\nLE8yMep65lqS7ZKckeT6JNcl2Xuk9XhPYXZJXgjcD5xaVc8ddT1zKckCYEFVXZZkW2AZcHBVzYtf\nkScJsE1V3Z9kC+Ai4F1VdfGIS5tTSf4CGAeeVFWvGnU9cynJzcB4Vc3LH68lWQpcWFUntqcun1BV\n94yqHs8U+lBVFwA/HXUdg1BVt1fVZW3+PuA6YMfRVjV3quf+trhFm+bVX0JJFgKvBE4cdS1aN0me\nDLwQOAmgqv7fKAMBDAVNkWQRsDtwyYhLmVPt0spyYBVwblXNq+8H/G/gfcDDI65jUAr4VpJlbZib\n+WRnYBL4bLv8d2KSbUZZkKEgAJI8ETgTeHdV/WzU9cylqnqoqhbT+4X8nknmzSXAJK8CVlXVslHX\nMkD7VtUe9EZPPrJdzp0vNgf2AE6oqt2BnwMjfWWAoSDatfYzgS9U1VdHXc+gtNPy84EDRlzKXNoH\n+JN23f004CVJPj/akuZWVa1sn6uAr9EbTXm+uA24bcrZ6xn0QmJkDIVNXLsRexJwXVV9ctT1zLUk\nY0m2a/NbAy8Frh9pUXOoqo6pqoVVtYjesDDfqao3jrisOZNkm/YABO2yysuAefMUYFX9BLg1ya6t\naX9G/KqADWqYiw1Vki8B+wHbJ7kN+GBVnTTaqubMPsCbgKvadXeAD1TVOaMraU4tAJa2Fzg9Dji9\nqubdY5vz2A7A13p/u7A58MWq+rfRljTn3gF8oT15dCNw2CiL8ZFUSVLHy0eSpI6hIEnqGAqSpI6h\nIEnqGAqSpI6hoI1Sko8meXGSg5Mcs47bjiW5pA0r8Efrefzd2qidlyd55gz9Dknyl+tzjDX2szjJ\ngX30G09y/GM9njZdhoI2VnsBFwMvAi5Yx233B66qqt2r6sJ1PXD7zcPBwBltH/85Q/dXAH09V59k\npt8NLQZmDYWqmqiqd/ZzPGk6hoI2Kkk+nuRK4A+A7wNvA05I8lfT9F2U5DtJrkxyXpJnJFkM/D1w\nUPtLf+s1ttm//fV/VXuPxpat/eYkf5fkMuD1wLuBt7d3UWyT5BvtnQ1XJ3l92yb0/jO/LMmHknwu\nyfeT/CjJ/2h99ktyYZKzgGvb+x8+245/eTsbejzwEeD1rebXt2Oe3N4VcXmSg6bs7+w2/6HW57tJ\nbkzyztY+bb0S+ItmbWSq6r1JTgfeDPwF8N2q2mct3f8RWFpVS5O8FTi+qg5uATJeVUdN7ZxkK+AU\nYP+q+mGSU4G30xuFFOCuNjAbSZ4F3F9V/5DkNcCPq+qVbd2TW//dgSuqqtovcn8PeAGwDXB5km+0\nfnsAz62qm5K8p/c163lJdgO+BTwL+I2ak/wtvSEt3tqG8bg0yben+TfYDXgxsC1wQ5IT6I39NF29\nkmcK2ijtAVxB7z+862botzfwxTb/OWDfWfa7K3BTVf2wLS+lN9b9al9ey3ZXAS9tZxJ/VFX3tvYD\ngH+d0u/rVfXL9rKY83lkYLdLq+qmNr8v8HmAqroeuIVeKKzpZcDRbWiS7wJbAc+Ypt83quqBdsxV\n9IaNWFu9kmcK2ni0Sz+n0BsC+07gCb3mLAf2rqpfDriEn0/X2M4q9qB3zf9vkpxXVR+h9x/3a6Z2\nXXPTmfY7iwCvqaobfqMx2WGNfg9MmX8I2HyGeiXPFLTxqKrl7b0IPwSeDXwHeHlVLV5LIPwHvZFD\nAf47MNtN5RuARUl+ty2/CfjebHUleTrwi6r6PPBxYI92SWbzqrprSteD2j2Dp9EbYPEH0+zuwlbr\n6ktUz2h13UfvEtBq3wTe0e5bkGT32eqcqd5+t9X855mCNipJxoC7q+rhJLvN8i7pd9B7o9V76b3d\nasbRJ6vqV0kOA77SngT6AfB/+ijrecDHkzwM/JrefYiXAmte47+S3mWj7YG/rqoft//4p/o0vRvn\nVwEPAm+pqgeSnM8jl4s+Cvw1vXsdVyZ5HHAT0O+7maerVwIcJVUaiCQnAidW1cVt+UO0G9MjLUya\nhWcK0gBU1dtGXYO0PjxTkCR1vNEsSeoYCpKkjqEgSeoYCpKkjqEgSer8fxv+1idmiTDEAAAAAElF\nTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# argument for removing scafs < 3 kb\n",
    "curated[\"orf_counts\"] = curated[\"scaffold\"].map(orf_counts)\n",
    "cgb = curated.query(\"scaffold_len<3000\").groupby(\"orf_counts\", as_index=False).aggregate(\"count\")\n",
    "sns.barplot(\"orf_counts\", \"manual\", data = cgb)\n",
    "plt.ylabel(\"# of contigs\")\n",
    "plt.xlabel(\"# of orfs/proteins\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### calculate rel. abund."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cmdir(udir + \"abund\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>scaffold_len</th>\n",
       "      <th>primary_cluster</th>\n",
       "      <th>secondary_cluster</th>\n",
       "      <th>prot_phylum_winner</th>\n",
       "      <th>prot_phylum_winner_perc</th>\n",
       "      <th>bin_classification</th>\n",
       "      <th>manual</th>\n",
       "      <th>reconciled_taxonomy</th>\n",
       "      <th>orf_counts</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OC1703_4500m_500m_sens_contig_3438907</td>\n",
       "      <td>11706</td>\n",
       "      <td>1</td>\n",
       "      <td>1_0</td>\n",
       "      <td>Chloroflexota</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Chloroflexota</td>\n",
       "      <td>12</td>\n",
       "      <td>OC1703_4500m_500m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OC1703_3000m_1000m_sens_contig_1343963</td>\n",
       "      <td>290156</td>\n",
       "      <td>2</td>\n",
       "      <td>2_1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>0.950758</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>264</td>\n",
       "      <td>OC1703_3000m_1000m</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 scaffold  scaffold_len primary_cluster  \\\n",
       "0   OC1703_4500m_500m_sens_contig_3438907         11706               1   \n",
       "1  OC1703_3000m_1000m_sens_contig_1343963        290156               2   \n",
       "\n",
       "  secondary_cluster prot_phylum_winner  prot_phylum_winner_perc  \\\n",
       "0               1_0      Chloroflexota                 0.833333   \n",
       "1               2_1     Proteobacteria                 0.950758   \n",
       "\n",
       "                                  bin_classification manual  \\\n",
       "0                                               None   None   \n",
       "1  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   None   \n",
       "\n",
       "  reconciled_taxonomy  orf_counts              sample  \n",
       "0       Chloroflexota          12   OC1703_4500m_500m  \n",
       "1      Proteobacteria         264  OC1703_3000m_1000m  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "working_set = curated.query(\"scaffold_len>3000\").query(\"manual!='eukaryote'\")\n",
    "working_set[\"sample\"] = working_set[\"scaffold\"].apply(lambda x: \"_\".join(x.split(\"_\")[0:3]))\n",
    "working_set.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# get coverage info \n",
    "with open(udir + \"/abund/runCoverM.sh\", \"w\") as outfile:\n",
    "    \n",
    "    for sample in working_set[\"sample\"].unique():\n",
    "        \n",
    "        sample_id = \"\".join(sample.split(\"_\")[1:3]) \n",
    "        # find self mapping\n",
    "        mapping = \"/scratch/groups/dekas/Nestor/new_ind_assembly/%s_mapped.bam\" %(sample_id)\n",
    "        sorted_bam = udir + \"abund/\" + sample_id + \"_mapped.sorted.bam\"\n",
    "        sort = \"samtools sort --threads 20 %s > %s\" %(mapping, sorted_bam)\n",
    "        index = \"samtools index -@ 20 %s\" %(sorted_bam)\n",
    "        call = \"coverm contig -t 20 -b %s --min-covered-fraction 0 --min-read-percent-identity 0.95 --output-format sparse -m count mean covered_fraction length > %s\" %(sorted_bam, \\\n",
    "            udir + \"abund/A\" + sample_id + \"_coverage_info.csv\")\n",
    "        outfile.write(sbatch(\"coverm\", 20, \" && \".join([sort, index, call])) + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>scaffold_len</th>\n",
       "      <th>primary_cluster</th>\n",
       "      <th>secondary_cluster</th>\n",
       "      <th>prot_phylum_winner</th>\n",
       "      <th>prot_phylum_winner_perc</th>\n",
       "      <th>bin_classification</th>\n",
       "      <th>manual</th>\n",
       "      <th>reconciled_taxonomy</th>\n",
       "      <th>orf_counts</th>\n",
       "      <th>sample</th>\n",
       "      <th>read_count</th>\n",
       "      <th>mean_coverage</th>\n",
       "      <th>breadth</th>\n",
       "      <th>total_read_count</th>\n",
       "      <th>total_coverage</th>\n",
       "      <th>read_fraction</th>\n",
       "      <th>coverage_fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OC1703_4500m_500m_sens_contig_3438907</td>\n",
       "      <td>11706</td>\n",
       "      <td>1</td>\n",
       "      <td>1_0</td>\n",
       "      <td>Chloroflexota</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Chloroflexota</td>\n",
       "      <td>12</td>\n",
       "      <td>OC1703_4500m_500m</td>\n",
       "      <td>723</td>\n",
       "      <td>8.810056</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>97378</td>\n",
       "      <td>2090.536231</td>\n",
       "      <td>0.007425</td>\n",
       "      <td>0.004214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OC1703_3000m_1000m_sens_contig_1343963</td>\n",
       "      <td>290156</td>\n",
       "      <td>2</td>\n",
       "      <td>2_1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>0.950758</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>264</td>\n",
       "      <td>OC1703_3000m_1000m</td>\n",
       "      <td>38955</td>\n",
       "      <td>18.916164</td>\n",
       "      <td>0.999697</td>\n",
       "      <td>558981</td>\n",
       "      <td>3078.515115</td>\n",
       "      <td>0.069689</td>\n",
       "      <td>0.006145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OC1703_3000m_3000m_sens_contig_701093</td>\n",
       "      <td>133857</td>\n",
       "      <td>2</td>\n",
       "      <td>2_1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>0.896825</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>126</td>\n",
       "      <td>OC1703_3000m_3000m</td>\n",
       "      <td>16587</td>\n",
       "      <td>17.429530</td>\n",
       "      <td>0.999873</td>\n",
       "      <td>313352</td>\n",
       "      <td>3891.659707</td>\n",
       "      <td>0.052934</td>\n",
       "      <td>0.004479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OC1703_4500m_2000m_sens_contig_1752643</td>\n",
       "      <td>19664</td>\n",
       "      <td>2</td>\n",
       "      <td>2_1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>0.956522</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>23</td>\n",
       "      <td>OC1703_4500m_2000m</td>\n",
       "      <td>1915</td>\n",
       "      <td>13.774572</td>\n",
       "      <td>0.998983</td>\n",
       "      <td>369313</td>\n",
       "      <td>4872.168753</td>\n",
       "      <td>0.005185</td>\n",
       "      <td>0.002827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OC1703_3000m_50m_sens_contig_1313614</td>\n",
       "      <td>9557</td>\n",
       "      <td>3</td>\n",
       "      <td>3_1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>12</td>\n",
       "      <td>OC1703_3000m_50m</td>\n",
       "      <td>1367</td>\n",
       "      <td>20.272139</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>584986</td>\n",
       "      <td>1204.464376</td>\n",
       "      <td>0.002337</td>\n",
       "      <td>0.016831</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 scaffold  scaffold_len primary_cluster  \\\n",
       "0   OC1703_4500m_500m_sens_contig_3438907         11706               1   \n",
       "1  OC1703_3000m_1000m_sens_contig_1343963        290156               2   \n",
       "2   OC1703_3000m_3000m_sens_contig_701093        133857               2   \n",
       "3  OC1703_4500m_2000m_sens_contig_1752643         19664               2   \n",
       "4    OC1703_3000m_50m_sens_contig_1313614          9557               3   \n",
       "\n",
       "  secondary_cluster prot_phylum_winner  prot_phylum_winner_perc  \\\n",
       "0               1_0      Chloroflexota                 0.833333   \n",
       "1               2_1     Proteobacteria                 0.950758   \n",
       "2               2_1     Proteobacteria                 0.896825   \n",
       "3               2_1     Proteobacteria                 0.956522   \n",
       "4               3_1     Proteobacteria                 1.000000   \n",
       "\n",
       "                                  bin_classification manual  \\\n",
       "0                                               None   None   \n",
       "1  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   None   \n",
       "2  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   None   \n",
       "3                                               None   None   \n",
       "4                                               None   None   \n",
       "\n",
       "  reconciled_taxonomy  orf_counts              sample  read_count  \\\n",
       "0       Chloroflexota          12   OC1703_4500m_500m         723   \n",
       "1      Proteobacteria         264  OC1703_3000m_1000m       38955   \n",
       "2      Proteobacteria         126  OC1703_3000m_3000m       16587   \n",
       "3      Proteobacteria          23  OC1703_4500m_2000m        1915   \n",
       "4      Proteobacteria          12    OC1703_3000m_50m        1367   \n",
       "\n",
       "   mean_coverage   breadth  total_read_count  total_coverage  read_fraction  \\\n",
       "0       8.810056  1.000000             97378     2090.536231       0.007425   \n",
       "1      18.916164  0.999697            558981     3078.515115       0.069689   \n",
       "2      17.429530  0.999873            313352     3891.659707       0.052934   \n",
       "3      13.774572  0.998983            369313     4872.168753       0.005185   \n",
       "4      20.272139  1.000000            584986     1204.464376       0.002337   \n",
       "\n",
       "   coverage_fraction  \n",
       "0           0.004214  \n",
       "1           0.006145  \n",
       "2           0.004479  \n",
       "3           0.002827  \n",
       "4           0.016831  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#assemblies\n",
    "amapping = pd.concat([pd.read_csv(item, sep=\"\\t\") for item \\\n",
    "                      in glob.glob(udir + \"abund/A*_info.csv\")]).rename(columns={\"Contig\": \"scaffold\", \"Read Count\": \"read_count\", \"Mean\": \"mean_coverage\", \"Covered Fraction\": \"breadth\"})\n",
    "working_set = working_set.merge(amapping[[\"scaffold\", \"read_count\", \"mean_coverage\", \"breadth\"]])\n",
    "wgb = working_set.query(\"breadth>0.50\").groupby([\"sample\"], as_index=False).aggregate({\"read_count\":\"sum\", \"mean_coverage\": \"sum\"}).rename(columns={\"read_count\": \"total_read_count\", \"mean_coverage\":\"total_coverage\"})\n",
    "working_set = working_set.query(\"breadth>0.50\").merge(wgb, how=\"left\")\n",
    "working_set[\"read_fraction\"] = working_set.apply(lambda x: (x[\"read_count\"]/x[\"total_read_count\"]), axis=1)\n",
    "working_set[\"coverage_fraction\"] = working_set.apply(lambda x: (x[\"mean_coverage\"]/x[\"total_coverage\"]), axis=1)\n",
    "working_set.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "working_set.to_csv(rootdir + \"tables/contig_working_set.tsv\", sep=\"\\t\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
